src/Microsoft.Health.Dicom.Core/Features/Store/DicomDatasetValidator.cs

```diff
@@ -90,22 +93,56 @@ string EnsureRequiredTagIsPresent(DicomTag dicomTag)
             }
             else
             {
-                ValidateIndexedItems(dicomDataset);
+                ValidateIndexedItems(dicomDataset, customTagEntries);
             }
         }
 
-        private void ValidateIndexedItems(DicomDataset dicomDataset)
+        private void ValidateIndexedItems(DicomDataset dicomDataset, IReadOnlyList<CustomTagEntry> customTagEntries)
         {
+            HashSet<DicomTag> standardTags = new HashSet<DicomTag>();
+            Dictionary<string, CustomTagEntry> privateTags = new Dictionary<string, CustomTagEntry>();
+            foreach (var customTagEntry in customTagEntries)
+            {
+                DicomTag dicomTag = DicomTag.Parse(customTagEntry.Path);
+                if (dicomTag.IsPrivate)
+                {
+                    privateTags.Add(customTagEntry.Path, customTagEntry);
+                }
+                else
+                {
+                    standardTags.Add(dicomTag);
+                }
+            }
+
+            // Process Standard Tag
             HashSet<DicomTag> indexableTags = QueryLimit.AllInstancesTags;
```

Comment:
`QueryLimit.AllInstanceTags` is mutable, as it is a plain ol' `HashSet<T>` (the `readonly` keyword only denotes that the reference itself is immutable). So the next `UnionWith` method is modifying the `static` `QueryLimit.AllInstancesTags` value!!

This is probably not what you want, right? Instead, we should make some changes:
1. Can `QueryLimit` be `internal`?
2. All of the `HashSet<T>` fields in `QueryLimit` should be:
    1. Either `IReadOnlyCollectiont<T>` or `ImmutableHashSet<T>` depending on your need
    2. Readonly properties (only `{ get; }`) instead of fields to make future changes easier

Reply:
I think the _public_ vs _internal_ is very common.